domain/relation/state/relation.go

```diff
@@ -481,6 +483,389 @@ WHERE  relation_id = $getRelation.relation_id
+) error {
+	// Check that the relation is a container scoped relation.
+	scope, err := st.getRelationScope(ctx, tx, relUUID)
+	if err != nil {
+		return errors.Errorf("getting relation scope: %w", err)
+	}
+	if scope != charm.ScopeContainer {
+		return nil
+	}
+
+	// Check that the other application in the relation is not a subordinate
+	// application, if it is, we have a relation between two subordinates, which
+	// is OK.
+	if subordinate, err := st.isSubordinate(ctx, tx, otherApplication); err != nil {
+		return errors.Errorf("checking if application is subordinate: %w", err)
+	} else if !subordinate {
+		return nil
+	}
+
+	// Check that the principle application of the unit is the other application
+	// in the relation the unit is trying to enter scope in.
+	principleAppID, err := st.getPrincipalApplicationOfUnit(ctx, tx, unitUUID)
+	if err != nil {
+		return errors.Errorf("getting principal application of unit: %w", err)
+	}
+	if principleAppID != otherApplication {
+		return errors.Errorf(
+			"unit cannot enter scope: principle application not in relation: %w",
+			relationerrors.PotentialRelationUnitNotValid,
+		)
+	}
+
+	return nil
+}
+
+// getRelationScope returns the scope of the given relation.
+func (st *State) getRelationScope(
+	ctx context.Context,
+	tx *sqlair.TX,
+	uuid corerelation.UUID,
+) (charm.RelationScope, error) {
+	type getScope struct {
+		Scope charm.RelationScope `db:"scope"`
+	}
+	relUUID := relationUUID{
+		UUID: uuid.String(),
+	}
+	var scope getScope
+	getScopeStmt, err := st.Prepare(`
+SELECT &getScope.*
+FROM   v_relation_endpoint
+WHERE  relation_uuid = $relationUUID.uuid
+`, relUUID, scope)
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+	err = tx.Query(ctx, getScopeStmt, relUUID).Get(&scope)
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+	return scope.Scope, nil
+}
+
+// isSubordinate returns true if the application is a subordinate application.
+func (st *State) isSubordinate(
+	ctx context.Context,
+	tx *sqlair.TX,
+	applicationUUID application.ID,
+) (bool, error) {
+	type getSubordinate struct {
+		ApplicationUUID application.ID `db:"application_uuid"`
+		Subordinate     bool           `db:"subordinate"`
+	}
+
+	subordinate := getSubordinate{
+		ApplicationUUID: applicationUUID,
+	}
+
+	stmt, err := st.Prepare(`
+SELECT subordinate AS &getSubordinate.subordinate
+FROM   application a
+JOIN   charm c ON a.charm_uuid = c.uuid
+JOIN   charm_metadata cm ON c.uuid = cm.charm_uuid
+WHERE  a.uuid = $getSubordinate.application_uuid
+`, subordinate)
+	if err != nil {
+		return false, errors.Capture(err)
+	}
+
+	err = tx.Query(ctx, stmt, subordinate).Get(&subordinate)
+	if err != nil {
+		return false, errors.Capture(err)
+	}
+
+	return subordinate.Subordinate, nil
+}
+
+// getPrincipalApplicationOfUnit returns the UUID of the principle application
+// of a unit. If the unit has no principle application, then the error
+// [relationerrors.UnitPrincipalNotFound] is returned.
```

Comment:
todo: add to the comment that the error is ambiguous if the unitUUID is not for a subordinate.  Or that it's assumed the unit is a subordinate. It feels like there are gotchas for future folks.

Reply:
Added a sentence to clarify that the error would always be returned if the unit is not a subordinate.